File: dongho-kim/2주차/Valid Parentheses/s.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();

        char[] chars = s.toCharArray();
        for (char ch : chars) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
                continue;
            }

            if (stack.isEmpty()) {
                return false;
            }

            char last = stack.peek();
            if (ch == ')' && last != '(') {
                return false;
            }
            if (ch == '}' && last != '{') {
                return false;
            }
            if (ch == ']' && last != '[') {
                return false;
            }
            stack.pop();
        }

        if (stack.isEmpty()) {
            return true;
        }
        return false;
    }
}

File: dongho-kim/2주차/Valid Parentheses/s.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
                continue

            if len(stack) == 0:
                return False

            if c == ')' and stack[-1] != '(':
                return False
            if c == '}' and stack[-1] != '{':
                return False
            if c == ']' and stack[-1] != '[':
                return False

            stack.pop()

        if len(stack) > 0:
            return False
        return True
